(PDB-154) allow zero gc interval

mullr opened the pull request.

ajroetker commented:
lgtm

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1868/

pljenkinsro commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1867/

wkalt commented on documentation/configure.markdown:
```diff
@@ -274,6 +274,8 @@ Do not use the `username` or `password` settings.
 
 This controls how often, in minutes, to compact the database. The compaction process reclaims space and deletes unnecessary rows. If not supplied, the default is every 60 minutes.
 
+If `gc-interval` is set to 0, all GC processes will be disabled. This may be useful, for example, for a read-only PuppetDB instance using a replicated database. It should not be used for a PuppetDB instance that will receive any commands. 
```
I don't think the part about receiving commands is accurate. The situation I'm familiar with is a single write database A shared by N PDB's dedicated to command processing, which each write to A and read off replicas of A. Under this scenario you may have a single PDB instance that does nothing but run GC on A, and receives no commands, but it's every instance _except_ that one on which you'd be setting gc-interval=0.

senior replied:
Seems like a more accurate way to put this is, if running more than one PuppetDB instance sharing a single database, only 1 of those instances should have GC enabled. Read vs. write vs. commands doesn't matter.

mullr replied:
I think receiving commands *is* what matters, since setting `gc-interval` to zero will also disable mq compaction.

wkalt commented on src/puppetlabs/puppetdb/cli/services.clj:
```diff
@@ -328,10 +328,11 @@
+          ;; Run database maintenance tasks seqentially to avoid
+          ;; competition. Each task must handle its own errors.
+          (gc-task db-maintenance-tasks)
+          (gc-task #(compress-dlo! dlo-compression-threshold discard-dir))))
```
this raises another question -- under the scenario I described above the DLO dirs that matter will be on the command processing slaves, despite the fact that the database we care about GCing is on a different host. Existing issue, but it seems like we need to decouple the DLO compression from the database cleanup.

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1876/

mullr commented:
@wkalt ok, added that feature

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1877/

ajroetker commented:
@pljenkinsro retest this please

pljenkinsro commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1886/

wkalt commented:
I think yesterday we were talking about either making the mq gc non-tunable or making it default to something else in the case of gc-interval=0, to avoid two flags. Is that accurate?

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/5/
Test FAILed.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-stable/4/
Test FAILed.

waynr commented:
test this please

pljenkinsro commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://jenkins.puppetlabs.com/job/platform_puppetdb_intn-sys_pr/1897/

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/6/
Test PASSed.

ajroetker commented:
So, reading @wkalt 's comment, it would seem that this PR is still in progress. Is that the case @mullr?

mullr commented:
@ajroetker Aye. Working on it now.

mullr commented:
Ok, I took out `disable-database-gc` and added `dlo-compression-interval`. By default this is equal to `gc-interval`, so it works the same for existing users. Changed the docs to tell people to set it if they want to zero-out `gc-interval`.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/21/
Test FAILed.

wkalt commented:
+1 from me. test failure looks transient.

ajroetker commented on src/puppetlabs/puppetdb/config.clj:
```diff
@@ -223,6 +225,11 @@
+                          ;; Default dlo-compression-interval to gc-interval
+                          (if (:dlo-compression-interval database-section)
+                            database-section
+                            (assoc database-section :dlo-compression-interval (:gc-interval database-section)))))
```
I think this is `assoc-when` or `update-when` from utils

mullr replied:
it's assoc-when-not, but I don't think that's a thing.

wkalt replied:
that's just assoc-when actually https://github.com/puppetlabs/puppetdb/blob/master/src/puppetlabs/puppetdb/utils.clj#L83-94

it's an implicit not I guess

mullr replied:
someone wrote assoc-when-not and called it assoc-when? *sigh*

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/34/
Test FAILed.

mullr commented:
@puppetlabs-jenkins retest this please

mullr commented:
test this please

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/35/
Test PASSed.

mullr commented:
@senior ⬆️

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/37/
Test FAILed.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/38/
Test PASSed.
